server/mailet/rate-limiter/src/main/scala/org/apache/james/transport/mailets/PerSenderRateLimit.scala

```diff
@@ -0,0 +1,220 @@
+      keyPrefix = keyPrefix,
+      entityType = Count)
+    recipientsRateLimiter = PerSenderRateLimiter(rateLimiter = extractRecipientsRules(duration).map(rateLimiterFactory.withSpecification),
+      keyPrefix = keyPrefix,
+      entityType = RecipientsType)
+    sizeRateLimiter = PerSenderRateLimiter(rateLimiter = extractSizeRules(duration).map(rateLimiterFactory.withSpecification),
+      keyPrefix = keyPrefix,
+      entityType = Size)
+    totalSizeRateLimiter = PerSenderRateLimiter(rateLimiter = extractTotalSizeRules(duration).map(rateLimiterFactory.withSpecification),
+      keyPrefix = keyPrefix,
+      entityType = TotalSize)
+  }
+
+  @VisibleForTesting
+  def parseDuration(): Duration =
+    Option(getInitParameter("duration"))
+      .map(string => DurationParser.parse(string, ChronoUnit.SECONDS))
+      .getOrElse(throw new IllegalArgumentException("'duration' is compulsory"))
+
+  private def extractCountRules(duration: Duration): Option[Rules] = Option(getInitParameter("count"))
+    .map(_.toLong)
+    .map(AllowedQuantity.liftOrThrow)
+    .map(quantity => Rules(Seq(Rule(quantity, duration))))
+
+  private def extractRecipientsRules(duration: Duration): Option[Rules] = Option(getInitParameter("recipients"))
+    .map(_.toLong)
+    .map(AllowedQuantity.liftOrThrow)
+    .map(quantity => Rules(Seq(Rule(quantity, duration))))
+
+  private def extractSizeRules(duration: Duration): Option[Rules] = Option(getInitParameter("size"))
+    .map {
+      case "" => throw new IllegalArgumentException("'size' field cannot be empty if specified")
+      case s => s
+    }
+    .map(org.apache.james.util.Size.parse)
+    .map(_.asBytes())
+    .map(AllowedQuantity.liftOrThrow)
+    .map(quantity => Rules(Seq(Rule(quantity, duration))))
+
+  private def extractTotalSizeRules(duration: Duration): Option[Rules] = Option(getInitParameter("totalSize"))
+    .map {
+      case "" => throw new IllegalArgumentException("'totalSize' field cannot be empty if specified")
+      case s => s
+    }
+    .map(org.apache.james.util.Size.parse)
+    .map(_.asBytes())
+    .map(AllowedQuantity.liftOrThrow)
+    .map(quantity => Rules(Seq(Rule(quantity, duration))))
+
+  override def service(mail: Mail): Unit = mail.getMaybeSender
+      .asOptional()
+      .ifPresent(sender => {
+        applyRateLimiter(mail, sender)
+      })
+
+  private def applyRateLimiter(mail: Mail, sender: MailAddress): Unit = {
+    val pivot: RateLimitingResult = AcceptableRate
+    val result = SFlux.merge(Seq(
+        countRateLimiter.rateLimit(sender, mail),
+        recipientsRateLimiter.rateLimit(sender, mail),
+        sizeRateLimiter.rateLimit(sender, mail),
+        totalSizeRateLimiter.rateLimit(sender, mail)))
+      .fold(pivot)((a, b) => a.merge(b))
+      .block()
+
+    if (result.equals(RateExceeded)) {
+      mail.setState(exceededProcessor)
```

Comment:
Before reading this line, 
I thought `exceededProcessor` will similar to a publisher, that sends events. (mechanism look like to MailboxListener).

Reply:
Processor is a logical groups of mailets and matchers, and it appears in the mailet container conf file.

This is a different domain area than Java publishers...